### ISSUE TITLE
GM_listValues shows lists values previously deleted with GM_deleteValue (only after page refresh)

### ISSUE THREAD

**Issue by Ryzhehvost** [Tampermonkey/tampermonkey#1635]
(Please fill out the issue template with your details)

## Expected Behavior

After script deletes some value with GM_deleteValue it should not be listed in GM_listValues (at least it was working like that recently)

## Actual Behavior

After script deletes some value with GM_deleteValue this value (correctly) not listed with GM_listValues until page refresh. After refresh it starts to list deleted value again, obviously when script tries to get it - result is `undefined`.

## Specifications

- both Firefox Developers Edition 107.0beta and Firefox 106.0.2
- TM:  4.18.0
- OS: Windows 10 20H2

## Script

Minimal reproducible example below. On first visit stores values, after first refresh they are removed and list is empty, on second refresh - list is back, but values are undefined.

```js
// ==UserScript==
// @name         New Userscript
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @author       You
// @match        *
// @icon         https://www.google.com/s2/favicons?sz=64&domain=github.com
// @grant        GM_deleteValue
// @grant        GM_setValue
// @grant        GM_getValue
// @grant        GM_listValues
// ==/UserScript==

(function() {
    'use strict';
    let values = GM_listValues();
    console.log(JSON.stringify(values));
    if (values.length == 0) {
    GM_setValue("test1","teststring1");
    GM_setValue("test2","teststring2");
    GM_setValue("test3","teststring3");
    } else {
        console.log("before:");
    values.forEach(function(key){
        console.log(key + " -> " + GM_getValue(key));
        GM_deleteValue(key);
    });
    }
    console.log("after:");
    let newvalues = GM_listValues();
    console.log(JSON.stringify(newvalues));
    newvalues.forEach(function(key){
        console.log(key + " -> " + GM_getValue(key));
    });

})();
```

**Comment by derjanb** [comment]
Nice finding. Thanks.

**Comment by derjanb** [comment]
Should be fixed at 4.19.6176 (in review|[crx](https://tampermonkey.net/crx/tampermonkey_beta_current.crx)|[xpi](https://firefox.tampermonkey.net/firefox-current-beta.xpi))

Please install the [BETA version](https://www.tampermonkey.net/index.php?browser=firefox) or check for BETA version updates at `about:addons` 

For a quick fix please export your settings and scripts as zip or (JSON) file at the "Utilities" tab and import it back at the fixed BETA version.

**Comment by TsXor** [comment]
> Should be fixed at 4.19.6176 (in review|[crx](https://tampermonkey.net/crx/tampermonkey_beta_current.crx)|[xpi](https://firefox.tampermonkey.net/firefox-current-beta.xpi))
> 
> Please install the [BETA version](https://www.tampermonkey.net/index.php?browser=firefox) or check for BETA version updates at `about:addons`
> 
> For a quick fix please export your settings and scripts as zip or (JSON) file at the "Utilities" tab and import it back at the fixed BETA version.

I encountered the same bug, and it is now fixed in BETA, thanks.

**Comment by ViperCode7** [comment]
I have something similar, but it's not fixed in 5.1.1.

I've opened 7 tabs in background, click by click, each put their URL in a key value pair, key being Date.now().
Then each tab checks to see if it's the last one added, if not it exists the code.
If it thinks it's the last one it continues to wait for another 5 seconds and puts all of those URLs into clipboard and removes them from storage. Or that's the idea.

For the love of me, I cannot figure out why tab 4 (just as example, it's random, sometimes it's tab 5 other times it's 6) would think it's the last one, when according to its time stamp, it's not, gets 4 or 5 values out of 7, while the rest exit. If I check the Storage tab in the script I can see the 7 key value pairs there before tab 4 does it's post 5 seconds task. Afterwards, 2 to 3 values remain in storage.

I do these checks in an interval, every 500ms, and the overall behavior is working, all except one tab exit the code. One tab remains to do the job, but that tab, even when I call GM_listValues every 500ms, even after 5 seconds have passed (also tried it with 10 seconds), it won't get all 7 values, just the random (4 or 5).

I don't want to open a tab and wait for it to finish loading until moving to the next one.

From what I can see, sometimes GM_listValues retrieves a limited list from storage, although the list is updated and full.

This would be the output of the working tab:
URL collection stopped!
 1729369894735 667 3
 1729369894735 1734 4
 1729369894735 2723 4
 1729369894735 3700 4
 1729369894735 4690 4
 1729369894735 5668 4
 1729369894735 6665 4
 1729369894735 7668 4
 1729369894735 8676 4
 1729369894735 9707 4
 1729369894735 10738 4
 prepping final task... 4
 clipboard set

As shown by the times this tab finishes setting up the storage, it's already populated by 3 other sources, then 4, but after that it stops, even if there were supposed to be 7. Meanwhile, in the script's storage tab, I repeatedly click reload and am able to see all 7 there.

Is it a limitations or something? Do only full page reloads get access the complete storage while pages already loaded access a shadow copy of it with actually incomplete data?

**Comment by derjanb** [comment]
@ViperCode7 Please create a new issue and add an example script. Thanks.

**Comment by ViperCode7** [comment]
It's fine... I take it TM has some parallelism limitations and precaching requirements that makes it unfeasible for a fix on my issue... I gave up on GM value and just used localStorage.
